kernel/src/main/scala/cats/kernel/instances/FunctionInstances.scala

```diff
@@ -106,11 +108,26 @@ private[instances] trait FunctionInstances4 {
     new Function1Semigroup[A, B] { def B: Semigroup[B] = S }
 }
 
+final private[instances] case class CombineFunction1[A, B](left: A => B, right: A => B, semiB: Semigroup[B])
+    extends (A => B) {
+  private def call(fn: A => B, a: A): TailRec[B] =
+    fn match {
+      case CombineFunction1(l, r, sg) =>
+        for {
+          lb <- tailcall(call(l, a))
+          rb <- tailcall(call(r, a))
+        } yield sg.combine(lb, rb)
+      case _ => done(fn(a))
+    }
+
+  final override def apply(a: A): B = call(this, a).result
+}
+
 trait Function1Semigroup[A, B] extends Semigroup[A => B] {
   implicit def B: Semigroup[B]
 
   override def combine(x: A => B, y: A => B): A => B =
-    (a: A) => B.combine(x(a), y(a))
+    CombineFunction1(x, y, B)
 }
```

Comment:
can we override `combineAllOption` which would actually be faster I think:
```scala
override def combineAllOption(fns: IterableOnce[A => B]): Option[A => B] =
  if (fns.isEmpty) None
  else Some { a: A =>
    B.combineAllOption(fns.map(_.apply(a))).get
  }
```

Reply:
Addressed